### PR TITLE
MAINT: Unify cached (C-level static) imports

### PR DESCRIPTION
**Pull request by seberg**
The main idea is that once we unify them, we can think about actually
being able to also clean them up.
This is far away from having no global state, but at this time I do not
see that as a seriousl goal personally. But it could be possible
to reload the module safely in a single interpreter maybe.

The commit adds the cache import for the complex casts, since it doubles
performance for small casts when the warning is not suppressed and it
unifies the handling.

---

Should be a trivial change, I did not think about using cached imports in places where we do not do really currently. My idea is, that unifying these, we could build a linked list with all cached entries and expose a function to clear/reset them. We probably could get reimports safe then (static interned strings are the main other thing, but at least they only leak a reference and don't really matter).

### CODE REVIEW

File: numpy/core/src/common/npy_import.h [hunk 1]
```diff
@@ -19,7 +19,7 @@
 NPY_INLINE static void
 npy_cache_import(const char *module, const char *attr, PyObject **cache)
 {
-    if (*cache == NULL) {
+    if (NPY_UNLIKELY(*cache == NULL)) {
```
Comment (seberg) [inline numpy/core/src/common/npy_import.h:22]: Can't hurt (unless we forgot the static somewhere ;)). Ideally, everything inside this if should probably be a non-inlined function at which point the `NPY_UNLIKELY` would not make a difference in any case (likely its not possible to measure it here already to be honest.)

**Review by eric-wieser — APPROVED** [review]
Nice clean up

**Comment by eric-wieser** [comment]
I suppose one path forward towards multiple interpreters would be to change the api to something like:
```C++
npy_cache_import(const char *module, const char *attr, PyObject *interp_state, npy_intp **cache_index, PyObject **cache) {
    if (cache_index != -1 && get_cache(interp_state)[cache_index] == *cache) {
        return;
    }
    *cache_index = get_cache_size(interp_state);
    realloc_cache_space_if_needed(interp_state, *cache_index + 1);
    *cache = get_cache(interp_state)[*cache_index] = get_the_module_attribute(module, attr);
}
```
This cache still avoids hitting the python C api, if you ignore the work needed to grab the `interp_state` object.

**Comment by seberg** [comment]
I think the solution in that case would be to add some slots to the module struct directly (no need for `interp_state`)? Since with subinterpreters you would need to create a new module object anyway. I assume there are solutions for this already, but I am not sure.

The bigger issue is for methods since they do not know the module, only the input object. And AFAIK Python has no story around how to make that work. For those you actually do need to get the interp state, and I am not sure how quick that is. I do not think Python actually uses this internally yet, so until then I am not sure I want to eve think about it.

The bigger issue is that we release and re-grab the GIL sometimes without knowing the GIL state. Until ufuncs and casting have proper error returns (which of course I want with new dtypes anyway, I admit). I do not think we can fully support sub-interpreters at all...

But yes, maybe it is less tricky then I thought, although there will be some old public API that can break it

**Comment by mattip** [comment]
> I do not think we can fully support sub-interpreters at all

This is my takeaway. We can make baby steps, but not any real progress. We should raise a nice error if we are reimported for now.

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**